odl/operator/pspace_ops.py

```diff
@@ -0,0 +1,189 @@
+                            ''.format(dom))
+        if ran is not None and not isinstance(ran, ProductSpace):
+            raise TypeError('space {!r} not a ProductSpace instance.'
+                            ''.format(dom))
+
+        # Convert ops to sparse representation
+        self.ops = sp.sparse.coo_matrix(operators)
+
+        if not all(isinstance(op, Operator) for op in self.ops.data):
+            raise TypeError('operators {!r} must be a matrix of operators.'
+                            ''.format(operators))
+
+        # Set domain and range (or verify if given)
+        if dom is None:
+            domains = [None] * self.ops.shape[1]
+
+        if ran is None:
+            ranges = [None] * self.ops.shape[0]
+
+        for row, col, op in zip(self.ops.row, self.ops.col, self.ops.data):
+            if domains[col] is None:
+                domains[col] = op.domain
+            elif domains[col] != op.domain:
+                raise ValueError('Column {}, has inconcistient domains,'
+                                 'got {} and {}'
+                                 ''.format(col, domains[col], op.domain))
+
+            if ranges[row] is None:
+                ranges[row] = op.range
+            elif ranges[row] != op.range:
+                raise ValueError('Row {}, has inconcistient ranges,'
+                                 'got {} and {}'
+                                 ''.format(row, ranges[row], op.range))
+
+        if dom is None:
+            for col in range(len(domains)):
+                if domains[col] is None:
+                    raise ValueError('Col {} empty, unable to determine '
+                                     'domain, please use dom parameter'
+                                     ''.format(col, domains[col]))
+
+            dom = ProductSpace(*domains)
+
+        if ran is None:
+            for row in range(len(ranges)):
+                if ranges[row] is None:
+                    raise ValueError('Row {} empty, unable to determine '
+                                     'range, please use ran parameter'
+                                     ''.format(row, ranges[row]))
+
+            ran = ProductSpace(*ranges)
+
+        # Set linearity
+        linear = all(op.is_linear for op in self.ops.data)
+
+        super().__init__(domain=dom, range=ran, linear=linear)
+
+    def __getitem__(self, *indices):
+        return self.ops__getitem__(*indices)
+
+    def _apply(self, x, out):
+        """ TODO
+
+        Parameters
+        ----------
+        todo
+
+        Returns
+        -------
+        None
+
+        Examples
+        --------
+        todo
+        """
+        has_evaluated_row = np.zeros(self.range.size, dtype=bool)
+        for i, j, op in zip(self.ops.row, self.ops.col, self.ops.data):
+            if not has_evaluated_row[i]:
+                op(x[j], out=out[i])
+            else:
+                # TODO: optimize
+                out[i] += op(x[j])
```

Comment:
Well that would be the best, a GEMM style call. But thats only useful in some cases, so maybe not worth it.

A more simple would be reusable temporaries.